### ISSUE TITLE
State type modified to generalized one when watch is called on generic type

### ISSUE THREAD

**Issue by vladimirkiva** [rrousselGit/river_pod#312]
In some cases (that are reproducibles) call to watch(provider.state) fires exception "type 'StateNotifierStateProvider<SuggestCompatibleState<User>>' is not a subtype of type 'StateNotifierStateProvider<UsersListState>' in type cast" (for the specified types).

The project that demostates the bug [https://github.com/vladimirkiva/riverpod_bug_q](https://github.com/vladimirkiva/riverpod_bug_q)

Code:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

// everything is simplified
abstract class SuggestCompatibleState<T> {
  int get count;
}

abstract class SuggestCompatibleModel<T, S extends SuggestCompatibleState<T>> extends StateNotifier<S> {
  SuggestCompatibleModel(SuggestCompatibleState<T> state) : super(state);

  void add();
}

class User {
}

class UsersListState extends SuggestCompatibleState<User> {
  int _counter = 0;
  Map<String, User> _users = Map();

  @override
  int get count => _users.length;

  UsersListState copy() {
    return UsersListState()
        .._counter = _counter
        .._users = _users
    ;
  }
}

class UsersModel extends SuggestCompatibleModel<User, UsersListState> {
  UsersModel() : super(UsersListState());

  @override
  void add() {
    state._users[state._users.length.toString()] = User();
    state = state;
  }

  int value() {
    return state.count;
  }
}

class Refresher extends StateNotifier<int> {
  Refresher() : super(0);

  void refresh() {
    state = state++;
  }
}

final usersProvider = StateNotifierProvider<UsersModel>((_) => UsersModel());
final refreshProvider = StateNotifierProvider<Refresher>((_) => Refresher());

void main() {
  runApp(
    ProviderScope(
      child: MyApp(),
    ),
  );
}

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends ConsumerWidget {
  @override
  Widget build(BuildContext context, watch) {
    final model = context.read(usersProvider);
    // !!!!!!!!!!!!!!!!!!!!!!!!! uncommenting this line forces to work without exception
    // watch(usersProvider.state);
    watch(refreshProvider.state);
    return Scaffold(
      appBar: AppBar(
        title: const Text('Demo'),
      ),
      body: Center(
        child: Column(
          children: [
            ElevatedButton(
                onPressed: context.read(usersProvider).add,
                child: Text('increment')
            ),
            ElevatedButton(onPressed: context.read(refreshProvider).refresh, child: Text('refresh page')),
            Text(model.value().toString()),
            GenericWidget(usersProvider),
            if (model.value() > 5) ...[
              DirectWidget()
            ],
            Text('Press "increment" more then 5 times (say, 7) and then press "refresh page" to get exception'),
          ],
        ),
      ),
    );
  }
}

class GenericWidget<T> extends StatelessWidget {
  final StateNotifierProvider<SuggestCompatibleModel<T, SuggestCompatibleState<T>>> _provider;

  const GenericWidget(this._provider) : super();

  @override
  Widget build(BuildContext _) {
      return Consumer(builder: (context, watch, __) {
        final SuggestCompatibleState<T> state = watch(_provider.state);
        return Text(state.count.toString());
      }
    );
  }
}

class DirectWidget extends ConsumerWidget {
  @override
  Widget build(BuildContext context, watch) {
    // !!!!!!!!!!!!!!!!!!!!!!!! Exception is thrown at this line
    // type 'StateNotifierStateProvider<SuggestCompatibleState<User>>' is not a subtype of type 'StateNotifierStateProvider<UsersListState>' in type cast
    final UsersListState state = watch(usersProvider.state);

    return Text(state.count.toString());
  }
}
```

To reproduce the bug:

1. run the application
2. press 'increment' more then 5 times
3. press 'refresh page' button

Exception should not be thrown. In case highlighted line is uncommented (line 84) the application works as expected.

**Comment by vladimirkiva** [comment]
Simplified test reproducing the bug:

```dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

// everything is simplified
abstract class SuggestCompatibleState<T> {
}

abstract class SuggestCompatibleModel<T, S extends SuggestCompatibleState<T>> extends StateNotifier<S> {
  SuggestCompatibleModel(SuggestCompatibleState<T> state) : super(state);
}

class User {
}

class UsersListState extends SuggestCompatibleState<User> {
}

class UsersModel extends SuggestCompatibleModel<User, UsersListState> {
  UsersModel() : super(UsersListState());
}

final usersProvider = StateNotifierProvider<UsersModel>((_) => UsersModel());

class MockWidget<T> {
  final StateNotifierProvider<SuggestCompatibleModel<T, SuggestCompatibleState<T>>> _provider;

  MockWidget(this._provider);

  read(ProviderContainer container) {
    container.read(_provider.state);
  }
}

void main() {
  test('Generic read', () {
    final container = ProviderContainer();
    final MockWidget w = MockWidget(usersProvider);
    w.read(container);
    // Exception is thrown here. Expected to pass without exception
    //  type 'StateNotifierStateProvider<SuggestCompatibleState<dynamic>>' is not a subtype of type 'StateNotifierStateProvider<UsersListState>' in type cast
    container.read(usersProvider.state);
  });
}
```

**Comment by rrousselGit** [comment]
It will likely require https://github.com/rrousselGit/river_pod/issues/341 to fix this

**Comment by vladimirkiva** [comment]
Not sure.

Here is even more simplified test:

```dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

abstract class SuggestCompatibleState<T> {
}

abstract class SuggestCompatibleModel<T, S extends SuggestCompatibleState<T>> extends StateNotifier<S> {
  SuggestCompatibleModel(SuggestCompatibleState<T> state) : super(state);
}

class User {
}

class UsersListState extends SuggestCompatibleState<User> {
}

class UsersModel extends SuggestCompatibleModel<User, UsersListState> {
  UsersModel() : super(UsersListState());
}

final usersProvider = StateNotifierProvider<UsersModel>((_) => UsersModel());

void main() {
  test('Runtime exception', () {
    final container = ProviderContainer();
    final StateNotifierProvider<SuggestCompatibleModel<User, SuggestCompatibleState<User>>> _provider = usersProvider;

    container.read(_provider.state);

    // Exception is thrown at the next line. Expected to pass without exception
    // type 'StateNotifierStateProvider<SuggestCompatibleState<dynamic>>' is not a subtype of type 'StateNotifierStateProvider<UsersListState>' in type cast
    container.read(usersProvider.state);
  });
}
```
There are two read operations. The second read generates runtime exception while everything compiles without errors.

**Comment by rrousselGit** [comment]
Yes but that's due to the ".state" relying on extensions to capture the type of a StateNotifier

It can't be fixed without changing the behaviour, so #341
